go/marshal/encode.go

```diff
@@ -223,18 +260,29 @@ func structEncoder(t reflect.Type, parentStructTypes []reflect.Type) encoderFunc
 
 	parentStructTypes = append(parentStructTypes, t)
 	fields, structType, originalFieldIndex := typeFields(t, parentStructTypes)
-	if structType != nil {
+
+	// It won't be possible to use the pre-computed struct type if it implements
+	// MarshalOptionsGetter, because the struct name can change at runtime.
+	isOptsGetter := t.Implements(optionsGetterType)
+
+	if originalFieldIndex != nil {
+		// Slowest path - we are extending some other struct. We need to start with the
```

Comment:
The problem is that if the orignal struct is named "A" and the options wants it to be named "B" the name will stay as "A" because with original we start with the original `types.Struct`.